File: Data Structures & Algorithms/largest-rectangle-in-histogram/submission-3.cpp
class Solution {
public:
    int largestRectangleArea(vector<int>& heights) {
        int res = 0;
        stack<pair<int, int>> tracker;
        for(int i = 0; i < heights.size(); i++){
            int start = i;
            while(!tracker.empty() && heights[i] < tracker.top().second){
                pair<int, int> point = tracker.top(); tracker.pop();
                int index = point.first;
                int height = point.second;
                res = max(res, (i - index) * height);
                start = index;
            }
            tracker.push({start, heights[i]});
        }

        while(!tracker.empty()){
            pair<int, int> point = tracker.top(); tracker.pop();
            int index = point.first;
            int height = point.second;
            res = max(res, ((int)heights.size() - index) * height);
        }
        return res;
    }
};

File: Data Structures & Algorithms/sliding-window-maximum/submission-3.py
class Solution:
    def maxSlidingWindow(self, nums: List[int], k: int) -> List[int]:
        res = []
        tracker = deque([])
        i = 0
        for j in range(len(nums)):
            while len(tracker) > 0 and tracker[0][0] < nums[j]:
                tracker.popleft()
            tracker.append((nums[j], j))
            while len(tracker) > 0 and tracker[-1][1] < i:
                tracker.pop()
            if j + 1 >= k:
                res.append(tracker[0][0])
                i += 1
        return res